src/services/epg.service.ts

```
@@ -0,0 +1,114 @@
+import { logDev } from '#src/utils/common';
+
+export const isFulfilled = <T>(input: PromiseSettledResult<T>): input is PromiseFulfilledResult<T> => input.status === 'fulfilled';
+
+export type EpgChannel = {
+  title: string;
+  image: string;
+  programs: EpgProgram[];
+};
+
+export type EpgProgram = {
+  id: string;
+  title: string;
+  startTime: string;
+  endTime: string;
+  description?: string;
+  image?: string;
+};
+
+const epgProgramSchema = object().shape({
+  id: string().required(),
+  title: string().required(),
+  startTime: string()
+    .required()
+    .test((value) => (value ? isValidDateString(value) : false)),
+  endTime: string()
+    .required()
+    .test((value) => (value ? isValidDateString(value) : false)),
+  chapterPointCustomProperties: array().of(
+    object().shape({
+      key: string().required(),
+      value: string().required(),
+    }),
+  ),
+});
+
+class EpgService {
+  async transformProgram(data: unknown) {
+    const incomingProgram = await epgProgramSchema.validate(data);
+    const transformed: EpgProgram = {
+      id: incomingProgram.id,
+      title: incomingProgram.title,
+      startTime: incomingProgram.startTime,
+      endTime: incomingProgram.endTime,
+    };
+
+    const customProperties = incomingProgram.chapterPointCustomProperties;
+
+    // map customProperties
+    if (customProperties) {
+      transformed.image = customProperties.find((item) => item.key === 'image')?.value || '';
+      transformed.description = customProperties.find((item) => item.key === 'description')?.value || '';
+    }
+
+    return transformed;
+  }
+
+  /**
+   * Ensure the given data validates to the EpgProgram schema
+   */
+  async parseSchedule(data: unknown) {
+    if (!Array.isArray(data)) return [];
+
+    const transformResults = await Promise.allSettled(data.map((program) => this.transformProgram(program)));
+
+    return transformResults.filter(isFulfilled).map((result) => result.value);
```

Looks like this will swallow any errors from invalid parsing. Should we rethrow an error if any are rejected?

---

Good point

---

I did this on purpose so that we only remove invalid programs from the schedule. The problem is that there are no indicators that there might be a problem. Do you have a suggestion?

---

I guess we should check with @marcovandeveen do decide how strict or forgiving we want to be with the parsing. If a program is invalid we have all the following options:
1. Silently drop it (better UX, but hard to debug)
2. Try to cast it as best we can (maybe most properties are valid and we can still use it, but might lead to unexpected results)
3. Show an error for the whole schedule (bad UX, but easier to debug)
4. Just show an error for the specific program (might be hard to display if we don't have valid start and end times for example.)

I think my votes would be for number 3 in dev and number 2 in prod.

---

Ok, we can still change this behavior if @marcovandeveen agrees

---

I'd also be fine with number 1 as is as long as we throw a logDev in for debugging

---

I actually had that coded, but removed it later 😅

---

Cool, let's go with that then unless Marco says otherwise